authority/provisioner/x5c.go

```diff
@@ -0,0 +1,264 @@
+type X5C struct {
+	Type      string `json:"type"`
+	Name      string `json:"name"`
+	Roots     string `json:"roots"`
+	rootPool  *x509.CertPool
+	Claims    *Claims `json:"claims,omitempty"`
+	claimer   *Claimer
+	audiences Audiences
+}
+
+// GetID returns the provisioner unique identifier. The name and credential id
+// should uniquely identify any X5C provisioner.
+func (p *X5C) GetID() string {
+	return "x5c/" + p.Name
+}
+
+// GetTokenID returns the identifier of the token.
+func (p *X5C) GetTokenID(ott string) (string, error) {
+	// Validate payload
+	token, err := jose.ParseSigned(ott)
+	if err != nil {
+		return "", errors.Wrap(err, "error parsing token")
+	}
+
+	// Get claims w/out verification. We need to look up the provisioner
+	// key in order to verify the claims and we need the issuer from the claims
+	// before we can look up the provisioner.
+	var claims jose.Claims
+	if err = token.UnsafeClaimsWithoutVerification(&claims); err != nil {
+		return "", errors.Wrap(err, "error verifying claims")
+	}
+	return claims.ID, nil
+}
+
+// GetName returns the name of the provisioner.
+func (p *X5C) GetName() string {
+	return p.Name
+}
+
+// GetType returns the type of provisioner.
+func (p *X5C) GetType() Type {
+	return TypeX5C
+}
+
+// GetEncryptedKey returns the base provisioner encrypted key if it's defined.
+func (p *X5C) GetEncryptedKey() (string, string, bool) {
+	return "", "", false
+}
+
+// Init initializes and validates the fields of a X5C type.
+func (p *X5C) Init(config Config) (err error) {
+	switch {
+	case p.Type == "":
+		return errors.New("provisioner type cannot be empty")
+	case p.Name == "":
+		return errors.New("provisioner name cannot be empty")
+	case p.Roots == "":
+		return errors.New("provisioner root(s) cannot be empty")
+	}
+
+	p.rootPool = x509.NewCertPool()
+	if len(p.Roots) > 0 && !p.rootPool.AppendCertsFromPEM([]byte(p.Roots)) {
+		return errors.Errorf("error parsing root certificate(s) for provisioner '%s'", p.Name)
+	}
+
+	// Update claims with global ones
+	if p.claimer, err = NewClaimer(p.Claims, config.Claims); err != nil {
+		return err
+	}
+
+	p.audiences = config.Audiences
+	return err
+}
+
+// authorizeToken performs common jwt authorization actions and returns the
+// claims for case specific downstream parsing.
+// e.g. a Sign request will auth/validate different fields than a Revoke request.
+func (p *X5C) authorizeToken(token string, audiences []string) (*x5cPayload, error) {
+	jwt, err := jose.ParseSigned(token)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error parsing token")
+	}
+
+	verifiedChains, err := jwt.Headers[0].Certificates(x509.VerifyOptions{
+		Roots: p.rootPool,
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "error verifying x5c certificate chain")
+	}
+	leaf := verifiedChains[0][0]
+
+	if leaf.KeyUsage&x509.KeyUsageDigitalSignature == 0 {
+		return nil, errors.New("certificate used to sign x5c token cannot be used for digital signature")
+	}
```

Comment:
Defining `KeyUsages` in `x509.VerifyOptions` does this for you, and I think it checks KeyUsage in the full chain not just in the leaf.

Reply:
Just checked and DigitalSignature is not an extended key usage. The [`KeyUsage`](https://godoc.org/crypto/x509#VerifyOptions) param appears to only accept [ExtKeyUsage](https://godoc.org/crypto/x509#ExtKeyUsage).

Reply:
Sounds good. But should the full chain support this key usage or just the leaf?

Reply:
Just as a point of reference, our root and intermediate certificates do not support digital signature extension. 

I think probably just the leaf needs to support it. @mmalone thoughts?

Reply:
sounds good